public/app/plugins/panel/annolist/AnnoListPanel.tsx

```diff
@@ -181,9 +181,22 @@ export class AnnoListPanel extends PureComponent<Props, State> {
   }
 
   onTagClick = (tag: string, remove?: boolean) => {
+    if (!remove && this.state.queryTags.includes(tag)) {
+      return;
+    }
+
     const queryTags = remove ? this.state.queryTags.filter((item) => item !== tag) : [...this.state.queryTags, tag];
 
-    this.setState({ queryTags });
+    // Logic to ensure keyboard focus isn't lost when the currently
+    // focused tag is removed
+    let nextTag: HTMLElement | undefined = undefined;
+    if (remove) {
+      const focusedTag = document.activeElement;
+      nextTag = (focusedTag!.parentElement?.nextElementSibling?.firstElementChild ??
+        focusedTag!.parentElement?.previousElementSibling?.firstElementChild) as HTMLElement;
+    }
```

Comment:
This seems pretty fragile to me - if the DOM structure changes this logic will become out of sync and fail in a pretty silent way? Especially considering (I think??) the DOM structure this is implying is of another component?

Perhaps another approach could be to just keep track of refs of the focusable elements, and move to the "next" item in that list when an item is deleted? Not sure...

Otherwise, some good tests here would help out.

Reply:
Changed so it's not so dependent on the DOM structure. Still not crazy about the solution, but perhaps it'll do until `AnnoListPanel` is changed to a functional component and we can use react-aria.